core/modules/czar/Czar.cc

```diff
@@ -0,0 +1,328 @@
+    if (hintIter != hints.end()) clientId = hintIter->second;
+
+    hintIter = hints.find("server_thread_id");
+    if (hintIter != hints.end()) {
+        try {
+            threadId = boost::lexical_cast<int>(hintIter->second);
+        } catch (boost::bad_lexical_cast const& exc) {
+            // Not fatal, just means we cannot associate query with particular
+            // client/thread and will not be able to kill it later
+        }
+    }
+
+    // this is atomic
+    uint64_t taskId = _idCounter++;
+    LOGF(_log, LOG_LVL_DEBUG, "taskId: %s" % taskId);
+
+    // make table names
+    auto taskIdStr = std::to_string(taskId);
+    std::string const resultName = _resultConfig.dbName + ".result_" + taskIdStr;
+    std::string const lockName = _resultConfig.dbName + ".message_" + taskIdStr;
+
+    std::vector<std::string> result(4);
+
+    // instantiate message table manager
+    MessageTable msgTable(lockName, _resultConfig);
+    try {
+        msgTable.lock();
+    } catch (std::exception const& exc) {
+        result[0] = exc.what();
+        return result;
+    }
+
+    // make new UserQuery
+    ccontrol::ConfigMap cm(hints);
+    std::string defDb = cm.get("db", "Failed to find default database, using empty string", "");
+    ccontrol::UserQuery::Ptr uq;
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+        uq = _uqFactory->newUserQuery(query, defDb, resultName, taskId);
+    }
+
+    // check for errors
+    auto error = uq->getError();
+    if (not error.empty()) {
+        result[0] = "Failed to instantiate query: " + error;
+        return result;
+    }
+
+    // start execution
+    LOGF(_log, LOG_LVL_DEBUG, "submitting new query");
+    uq->submit();
+
+    // spawn background thread to wait until query finishes to unlock
+    auto finalizer = [](ccontrol::UserQuery::Ptr uq, MessageTable msgTable) {
+        uq->join();
+        try {
+            msgTable.unlock(uq);
+            if (uq) uq->discard();
+        } catch (std::exception const& exc) {
+            // TODO? if this fails there is no way to notify client, and client
+            // will likely hang because table may still be locked.
+        }
+    };
+    LOGF(_log, LOG_LVL_DEBUG, "starting finalizer thread for query");
+    std::thread finalThread(finalizer, uq, msgTable);
+    finalThread.detach();
+
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+
+        // first cleanup client query map from completed queries
+        for (auto iter = _clientToQuery.begin(); iter != _clientToQuery.end(); ) {
+            if (iter->second.expired()) {
+                iter = _clientToQuery.erase(iter);
+            } else {
+                ++ iter;
+            }
+        }
+
+        // remember query (weak pointer) in case we want to kill query
+        // TODO: the map grows indefinitely now, will have to do some cleanup
+        if (not clientId.empty() and threadId >= 0) {
+            ClientThreadId ctId(clientId, threadId);
```

Comment:
This needs some thought, right now it is just a transparent rewrite of the XML-RPC handling of this. In principle you are correct that `threadId` is probably sufficient to identify client in current system, but this depends on your knowledge that "client_dst_name" is never going to change which is a sort of implementation detail. 
In the long term I think we should switch from this czar-internal mapping to using QMeta for this info, I just did not want to do it in this ticket. So let's keep this as it is now and we'll come back to it later anyway when we start implementing async queries.